### PR TITLE
TISPARK-450: Fix tiflash task distributes as tikv

### PR DESCRIPTION
**Pull request by zanmato1984**
### What problem does this PR solve? 
https://internal.pingcap.net/jira/browse/TISPARK-450

### What is changed and how it works?
1. Add new structure `TiStoreType` identifying which store (TiKV/TiFlash) to read;
2. Let region task generation (by `RegionSplitter`) get store based on `TiStoreType`;
3. Point 2 also handles retrying on TiFlash region movement, as region task regeneration is still based on `TiStoreType`;
4. Hard code TiFlash engine label as every other project (TiKV/TiFlash/TiDB) does;
5. Remove some useless methods.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java [hunk 1]
```diff
@@ -184,26 +185,41 @@ public TiRegion getRegionById(long regionId) {
...
-    long storeId = leader.getStoreId();
-    return Pair.create(region, cache.getStoreById(storeId));
-  }
 
-  public Pair<TiRegion, Store> getRegionStorePairByRegionId(long id) {
-    TiRegion region = cache.getRegionById(id);
-    if (!region.isValid()) {
-      throw new TiClientInternalException("Region invalid: " + region.toString());
+    Store store = null;
+    if (storeType == null || storeType == TiStoreType.TiKV) {
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java]: when is it possible that `storeType` is null?

Comment (zanmato1984) [reply tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java]: Never. This is one thing I missed after several round of refining TiStoreType structure. Addressed.

File: tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java [hunk 2]
```diff
@@ -184,26 +185,41 @@ public TiRegion getRegionById(long regionId) {
...
+      Peer leader = region.getLeader();
+      store = cache.getStoreById(leader.getStoreId());
+    } else {
+      for (Peer peer : region.getLearnerList()) {
+        Store s = getStoreById(peer.getStoreId());
+        for (Metapb.StoreLabel label : s.getLabelsList()) {
+          if (label.getKey().equals(storeType.getLabelKey())
+              && label.getValue().equals(storeType.getLabelValue())) {
+            store = s;
+            break;
```
Comment (zhexuany) [inline tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java]: this `break` only breaks inner loop. Maybe you can use the following code 

```
outerloop:
 for (Peer peer : region.getLearnerList()) {
         Store s = getStoreById(peer.getStoreId());
         for (Metapb.StoreLabel label : s.getLabelsList()) {
           if (label.getKey().equals(storeType.getLabelKey())
               && label.getValue().equals(storeType.getLabelValue())) {
             store = s;
             break outerloop;
            }
         }
   }

```

Comment (zanmato1984) [reply tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java]: Nice catch. Addressed.

**Pull request merged**